// erxes-common-ui/lib/components/helpers/typography.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fontWeightLight = 300;
var fontWeightRegular = 400;
var fontWeightMedium = 500;
var fontSizeHeading1 = 64;
var fontSizeHeading2 = 48;
var fontSizeHeading3 = 40;
var fontSizeHeading4 = 32;
var fontSizeHeading5 = 24;
var fontSizeHeading6 = 20;
var fontSizeHeading7 = 16;
var fontSizeHeading8 = 12;
var fontSizeBody = 13;
var fontSizeUppercase = 11;
var lineHeightHeading1 = 72 / fontSizeHeading1;
var lineHeightHeading2 = 56 / fontSizeHeading2;
var lineHeightHeading3 = 48 / fontSizeHeading3;
var lineHeightHeading4 = 40 / fontSizeHeading4;
var lineHeightHeading5 = 32 / fontSizeHeading5;
var lineHeightHeading6 = 28 / fontSizeHeading6;
var lineHeightHeading7 = 24 / fontSizeHeading7;
var lineHeightHeading8 = 16 / fontSizeHeading8;
var lineHeightBody = 20 / fontSizeBody;
var lineHeightUppercase = 11 / fontSizeUppercase;
exports.default = {
    fontWeightLight: fontWeightLight,
    fontWeightRegular: fontWeightRegular,
    fontWeightMedium: fontWeightMedium,
    fontSizeHeading1: fontSizeHeading1,
    fontSizeHeading2: fontSizeHeading2,
    fontSizeHeading3: fontSizeHeading3,
    fontSizeHeading4: fontSizeHeading4,
    fontSizeHeading5: fontSizeHeading5,
    fontSizeHeading6: fontSizeHeading6,
    fontSizeHeading7: fontSizeHeading7,
    fontSizeHeading8: fontSizeHeading8,
    fontSizeBody: fontSizeBody,
    fontSizeUppercase: fontSizeUppercase,
    lineHeightHeading1: lineHeightHeading1,
    lineHeightHeading2: lineHeightHeading2,
    lineHeightHeading3: lineHeightHeading3,
    lineHeightHeading4: lineHeightHeading4,
    lineHeightHeading5: lineHeightHeading5,
    lineHeightHeading6: lineHeightHeading6,
    lineHeightHeading7: lineHeightHeading7,
    lineHeightHeading8: lineHeightHeading8,
    lineHeightBody: lineHeightBody,
    lineHeightUppercase: lineHeightUppercase
};
