// erxes-common-ui/lib/components/colors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("./color");
// Core colors
var colorPrimary = '#673FBD';
var colorPrimaryDark = '#5629B6';
var colorSecondary = '#6569DF';
var colorCoreRed = '#EA475D';
var colorCoreTeal = '#63D2D6';
var colorCoreYellow = '#F7CE53';
var colorCoreOrange = '#FF6600';
var colorCoreGreen = '#3CCC38';
var colorCoreBlue = '#3B85F4';
var colorCoreDarkBlue = '#0a1e41';
var colorCoreBlack = '#393C40';
var colorCoreGray = '#888';
var colorCoreLightGray = '#AAAEB3';
var colorLightGray = '#AAA';
var colorLightBlue = '#F8FBFF';
var colorCoreDarkGray = '#373737';
var colorShadowGray = '#DDD';
var colorBlack = '#000';
var colorWhite = '#FFF';
// backgrounds
var bgMain = '#EDF1F5';
var bgDark = color_1.rgba(colorBlack, 0.95);
var bgLight = '#FAFAFA';
var bgActive = '#F0F0F0';
var bgGray = '#e6e6e6';
var bgLightPurple = '#F7F8FC';
var bgUnread = '#ededfb';
var bgInternal = '#FFFCCC';
// Link colors
var linkPrimary = '#1785FC';
var linkPrimaryHover = color_1.rgba(linkPrimary, 0.7);
// Border colors
var borderPrimary = '#EEE';
var borderDarker = '#DEE4E7';
// Text colors
var textPrimary = '#444';
var textSecondary = color_1.rgba(textPrimary, 0.8);
// Shadow colors
var shadowPrimary = color_1.rgba(colorBlack, 0.08);
var darkShadow = color_1.rgba(colorBlack, 0.2);
// Social colors
var socialFacebook = '#3A5999';
var socialFacebookMessenger = '#1472FB';
var socialTwitter = '#1DA1F2';
var socialGmail = '#D44638';
var socialGoogleMeet = '#038476';
var socialWhatsApp = '#25D366';
var socialTelegram = '#0088cc';
var socialViber = '#8f5db7';
var socialLine = '#00c300';
var socialTwilio = '#cf272d';
exports.default = {
    colorPrimary: colorPrimary,
    colorPrimaryDark: colorPrimaryDark,
    colorSecondary: colorSecondary,
    colorCoreRed: colorCoreRed,
    colorCoreTeal: colorCoreTeal,
    colorCoreYellow: colorCoreYellow,
    colorCoreOrange: colorCoreOrange,
    colorCoreBlue: colorCoreBlue,
    colorCoreGreen: colorCoreGreen,
    colorCoreBlack: colorCoreBlack,
    colorCoreGray: colorCoreGray,
    colorCoreLightGray: colorCoreLightGray,
    colorWhite: colorWhite,
    colorBlack: colorBlack,
    colorShadowGray: colorShadowGray,
    colorLightGray: colorLightGray,
    colorLightBlue: colorLightBlue,
    colorCoreDarkGray: colorCoreDarkGray,
    colorCoreDarkBlue: colorCoreDarkBlue,
    bgMain: bgMain,
    bgDark: bgDark,
    bgLight: bgLight,
    bgActive: bgActive,
    bgUnread: bgUnread,
    bgLightPurple: bgLightPurple,
    bgInternal: bgInternal,
    bgGray: bgGray,
    linkPrimary: linkPrimary,
    linkPrimaryHover: linkPrimaryHover,
    borderPrimary: borderPrimary,
    borderDarker: borderDarker,
    textPrimary: textPrimary,
    textSecondary: textSecondary,
    shadowPrimary: shadowPrimary,
    darkShadow: darkShadow,
    socialFacebook: socialFacebook,
    socialFacebookMessenger: socialFacebookMessenger,
    socialTwitter: socialTwitter,
    socialGmail: socialGmail,
    socialGoogleMeet: socialGoogleMeet,
    socialTelegram: socialTelegram,
    socialViber: socialViber,
    socialLine: socialLine,
    socialTwilio: socialTwilio,
    socialWhatsApp: socialWhatsApp
};
