// erxes-common-ui/lib/components/button/index.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var styled_components_1 = __importStar(require("styled-components"));
var styled_components_ts_1 = __importDefault(require("styled-components-ts"));
var icon_1 = __importDefault(require("../icon"));
var colors_1 = __importDefault(require("../colors"));
var color_1 = require("../colors/color");
var types = {
    default: {
        background: colors_1.default.colorPrimary
    },
    primary: {
        background: colors_1.default.colorSecondary,
        border: color_1.darken(colors_1.default.colorSecondary, 20)
    },
    success: {
        background: colors_1.default.colorCoreGreen
    },
    danger: {
        background: colors_1.default.colorCoreRed
    },
    warning: {
        background: colors_1.default.colorCoreYellow,
        border: color_1.darken(colors_1.default.colorCoreYellow, 25)
    },
    simple: {
        background: 'rgba(0,0,0,0.05)',
        color: colors_1.default.colorCoreGray,
        border: colors_1.default.colorCoreGray
    },
    link: {
        background: 'transparent',
        color: colors_1.default.colorCoreGray
    }
};
var sizes = {
    large: {
        padding: '10px 30px',
        fontSize: '13px'
    },
    medium: {
        padding: '7px 20px',
        fontSize: '12px'
    },
    small: {
        padding: '5px 15px',
        fontSize: '10px'
    }
};
var ButtonStyled = styled_components_ts_1.default(styled_components_1.default.button)(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  border-radius: 30px;\n  position: relative;\n  transition: all 0.3s ease;\n  outline: 0;\n\n  ", ";\n\n  a {\n    color: ", ";\n  }\n\n  & + button,\n  + a,\n  + span {\n    margin-left: 10px;\n  }\n\n  > i + span,\n  > span + i,\n  > span i {\n    margin-left: 5px;\n  }\n"], ["\n  border-radius: 30px;\n  position: relative;\n  transition: all 0.3s ease;\n  outline: 0;\n\n  ",
    ";\n\n  a {\n    color: ", ";\n  }\n\n  & + button,\n  + a,\n  + span {\n    margin-left: 10px;\n  }\n\n  > i + span,\n  > span + i,\n  > span i {\n    margin-left: 5px;\n  }\n"])), function (props) { return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    padding: ", ";\n    background: ", ";\n    font-size: ", ";\n    text-transform: ", ";\n    color: ", " !important;\n    border: none;\n    display: ", ";\n    width: ", ";\n    font-weight: ", ";\n\n    &:hover {\n      cursor: pointer;\n      text-decoration: none;\n      color: ", ";\n      background: ", ";\n    }\n\n    &:active,\n    &:focus {\n      box-shadow: ", ";\n      box-shadow: ", ";\n    }\n\n    &:disabled {\n      cursor: not-allowed !important;\n      opacity: 0.75;\n    }\n  "], ["\n    padding: ", ";\n    background: ", ";\n    font-size: ",
    ";\n    text-transform: ", ";\n    color: ",
    " !important;\n    border: none;\n    display: ", ";\n    width: ", ";\n    font-weight: ", ";\n\n    &:hover {\n      cursor: pointer;\n      text-decoration: none;\n      color: ",
    ";\n      background: ",
    ";\n    }\n\n    &:active,\n    &:focus {\n      box-shadow: ",
    ";\n      box-shadow: ", ";\n    }\n\n    &:disabled {\n      cursor: not-allowed !important;\n      opacity: 0.75;\n    }\n  "])), sizes[props.hugeness].padding, types[props.btnStyle].background, props.uppercase
    ? sizes[props.hugeness].fontSize
    : "calc(" + sizes[props.hugeness].fontSize + " + 1px)", props.uppercase ? 'uppercase' : 'none', types[props.btnStyle].color
    ? types[props.btnStyle].color
    : colors_1.default.colorWhite, props.block && 'block', props.block && '100%', !props.uppercase && '500', types[props.btnStyle].color &&
    color_1.darken(types[props.btnStyle].color, 35), props.btnStyle !== 'link' &&
    "" + color_1.darken(types[props.btnStyle].background, 20), types[props.btnStyle].border
    ? "0 0 0 0.2rem " + color_1.lighten(types[props.btnStyle].border, 65)
    : "0 0 0 0.2rem " + color_1.lighten(types[props.btnStyle].background, 65), props.btnStyle === 'link' && 'none'); }, colors_1.default.colorWhite);
var ButtonLink = styled_components_ts_1.default(styled_components_1.default(ButtonStyled.withComponent('a')))(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  text-decoration: inherit;\n  text-align: center;\n  \n  ", ";\n"], ["\n  text-decoration: inherit;\n  text-align: center;\n  \n  ",
    ";\n"])), function (props) {
    return props.disabled && styled_components_1.css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n      cursor: not-allowed !important;\n      opacity: 0.75;\n\n      &:focus {\n        text-decoration: none;\n      }\n    "], ["\n      cursor: not-allowed !important;\n      opacity: 0.75;\n\n      &:focus {\n        text-decoration: none;\n      }\n    "])));
});
var ButtonGroup = styled_components_ts_1.default(styled_components_1.default.div)(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  position: relative;\n\n  button + a,\n  a + button {\n    margin-left: ", ";\n  }\n\n  ", ";\n"], ["\n  position: relative;\n\n  button + a,\n  a + button {\n    margin-left: ", ";\n  }\n\n  ",
    ";\n"])), function (props) { return props.hasGap && '10px'; }, function (props) {
    return !props.hasGap && styled_components_1.css(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n      button,\n      a {\n        margin: 0;\n      }\n\n      > button:not(:last-child),\n      > a:not(:last-child) {\n        border-top-right-radius: 0;\n        border-bottom-right-radius: 0;\n        border-right: 1px solid rgba(0, 0, 0, 0.13);\n      }\n\n      > button:not(:first-child),\n      > a:not(:first-child) {\n        border-top-left-radius: 0;\n        border-bottom-left-radius: 0;\n      }\n    "], ["\n      button,\n      a {\n        margin: 0;\n      }\n\n      > button:not(:last-child),\n      > a:not(:last-child) {\n        border-top-right-radius: 0;\n        border-bottom-right-radius: 0;\n        border-right: 1px solid rgba(0, 0, 0, 0.13);\n      }\n\n      > button:not(:first-child),\n      > a:not(:first-child) {\n        border-top-left-radius: 0;\n        border-bottom-left-radius: 0;\n      }\n    "])));
});
var Button = /** @class */ (function (_super) {
    __extends(Button, _super);
    function Button() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Button.prototype.render = function () {
        var _a = this.props, _b = _a.size, size = _b === void 0 ? 'medium' : _b, _c = _a.btnStyle, btnStyle = _c === void 0 ? 'default' : _c, _d = _a.block, block = _d === void 0 ? false : _d, _e = _a.type, type = _e === void 0 ? 'button' : _e, _f = _a.uppercase, uppercase = _f === void 0 ? true : _f, commonExcluded = __rest(_a, ["size", "btnStyle", "block", "type", "uppercase"]);
        var href = commonExcluded.href, children = commonExcluded.children, icon = commonExcluded.icon;
        var props = __assign({}, commonExcluded, { hugeness: size, btnStyle: btnStyle, block: block, type: type, uppercase: uppercase });
        var Element = href ? ButtonLink : ButtonStyled;
        if (icon) {
            return (react_1.default.createElement(Element, __assign({}, props),
                react_1.default.createElement(icon_1.default, { icon: icon }),
                children && react_1.default.createElement("span", null, children)));
        }
        return react_1.default.createElement(Element, __assign({}, props), children);
    };
    Button.Group = Group;
    Button.defaultProps = {
        btnStyle: 'default',
        size: 'medium',
        block: false,
        type: 'button',
        uppercase: true
    };
    return Button;
}(react_1.default.Component));
function Group(_a) {
    var children = _a.children, _b = _a.hasGap, hasGap = _b === void 0 ? true : _b;
    return react_1.default.createElement(ButtonGroup, { hasGap: hasGap }, children);
}
exports.default = Button;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
