// erxes-common-ui/lib/components/form/DateControl.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_datetime_1 = __importDefault(require("@nateradebaugh/react-datetime"));
var react_1 = __importDefault(require("react"));
var styles_1 = require("./styles");
var DateControl = /** @class */ (function (_super) {
    __extends(DateControl, _super);
    function DateControl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DateControl.prototype.componentDidMount = function () {
        var registerChild = this.props.registerChild;
        if (registerChild) {
            registerChild(this);
        }
    };
    DateControl.prototype.render = function () {
        var _this = this;
        var _a = this.props, errors = _a.errors, value = _a.value, name = _a.name, placeholder = _a.placeholder, dateFormat = _a.dateFormat, timeFormat = _a.timeFormat, required = _a.required;
        var errorMessage = errors && errors[name || ''];
        // cancel custom browser default form validation error
        var onChange = function (e) {
            if (_this.props.onChange) {
                _this.props.onChange(e);
            }
        };
        var inputProps = {
            name: name,
            placeholder: placeholder || '',
            required: required || false,
            autoComplete: 'off'
        };
        var attributes = {
            inputProps: inputProps,
            dateFormat: dateFormat,
            timeFormat: timeFormat || false,
            value: value,
            closeOnSelect: true,
            onChange: onChange,
            utc: true
        };
        return (react_1.default.createElement(styles_1.FlexWrapper, null,
            react_1.default.createElement(react_datetime_1.default, __assign({}, attributes)),
            errorMessage));
    };
    DateControl.defaultProps = {
        dateFormat: 'MMM,DD YYYY'
    };
    return DateControl;
}(react_1.default.Component));
exports.default = DateControl;
